[lldb/Interpreter] Introduce `ScriptedStopHook{,Python}Interface` & make use of it

medismailben opened the pull request:
This patch introduces new `ScriptedStopHook{,Python}Interface` classes that make use of the Scripted Interface infrastructure and makes use of it in `StopHookScripted`.

It also relax the requirement on the number of argument for initializing scripting extension if the size of the interface parameter pack contains 1 less element than the extension maximum number of positional arguments for this initializer.
This addresses the cases where the embedded interpreter session dictionary is passed to the extension initializer which is not used most of the time.

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -137,12 +137,28 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
-            expected_return_object = init(args...);
-          },
-          transformed_args);
+      size_t num_args = sizeof...(Args);
+      if (num_args != arg_info->max_positional_args) {
+        if (num_args != arg_info->max_positional_args - 1)
+          return create_error(
+              llvm::formatv("Passed arguments ({0}) doesn't match the number "
+                            "of expected arguments ({1}).",
+                            num_args, arg_info->max_positional_args));
+
+        std::apply(
+            [&init, &expected_return_object](auto &&...args) {
+              llvm::consumeError(expected_return_object.takeError());
+              expected_return_object = init(args...);
+            },
+            std::tuple_cat(transformed_args, std::make_tuple(dict)));
+      } else {
+        std::apply(
+            [&init, &expected_return_object](auto &&...args) {
+              llvm::consumeError(expected_return_object.takeError());
+              expected_return_object = init(args...);
+            },
+            transformed_args);
+      }
```
Your commit message explains it, but I think this deserves a comment here as well.

JDevlieghere reviewed (approved):
LGTM!

medismailben commented:
Depends on https://github.com/llvm/llvm-project/pull/109063/files

medismailben committed:
[lldb/Interpreter] Relax Scripted Interface object creation requirements

Historically, some scripted extension would pass the embedded
interpreter session dictionary as a argument of the extension
initializer.

In many cases, that argument ended up not being used but was still
required for all new extension implementations.

This patch relax this requirement by checking if the number of maximum
positional argument is 1 argument more then the number of argument
passed in the parameter pack, in which case it will pass the interpreter
session dictionary with the rest of the arguments. If the implementation
doesn't have it in the initializer, only the paramater pack is passed to it.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben committed:
[lldb/Interpreter] Make use of ScriptedStopHook{,Python}Interface

This patch makes use of the previously defined
`ScriptedStopHook{,Python}Interface` in the `StopHookScripted` class and
removes all the SWIG methods that were implemented specifically for this.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben committed:
[lldb/Interpreter] Create ScriptedStopHook{,Python}Interface (NFC)

This patch introduces a new scripted interface for ScriptedStopHooks.

This will be used in a follow-up patch to call into the script methods.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

DavidSpickett commented on lldb/unittests/ScriptInterpreter/Python/PythonTestSuite.cpp:
```diff
@@ -269,21 +274,7 @@ void *lldb_private::python::SWIGBridge::LLDBSWIGPython_GetDynamicSetting(
-}
-
-python::PythonObject
-lldb_private::python::SWIGBridge::ToSWIGWrapper(Status status) {
```
FYI https://github.com/llvm/llvm-project/commit/76b827bb4d5b4cc4d3229c4c6de2529e8b156810

DavidSpickett commented:
I've reverted this (https://github.com/llvm/llvm-project/commit/801046e3303eed43bffebb84e9e505cc19cad5c0) because it broke a stop hook test:
https://lab.llvm.org/buildbot/#/builders/162/builds/6735

It seems to add the stop hook but there's no output when it should have been called.
